### ISSUE TITLE
[FR] improve _generate_square_subsequent_mask function in nn.Transformer

### ISSUE THREAD

**Issue by Kyeongpil** [pytorch/pytorch#48360]
## 🚀 Feature
make _generate_square_subsequent_mask function a little bit more efficient.

## Motivation
_generate_square_subsequent_mask function in nn.Transformer uses torch.triu and transpose() function in order to create a lower triangular matrix of attention mask.
However, I think it would be efficient if the function just calls torch.tril function.

## Pitch

```
mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
```
in _generate_square_subsequent_mask can be change to 
```
mask = torch.tril(torch.ones(sz, sz)) == 1
mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
```
which doesn't need to use transpose function

## Alternatives
I think it also can be replaced to
```
mask = torch.tril(torch.ones(sz, sz))
mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
```

## Additional context
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L133

cc @zhangguanheng66

**Comment by bhosmer** [comment]
Fixed in #60631 - @Kyeongpil thanks for calling this out!

**Comment by Kyeongpil** [comment]
@bhosmer Thank you for considering this feature request :)
